From 7e83e5a090150932056a36aea11c8f33529f1e0e Mon Sep 17 00:00:00 2001
From: Kreijstal <elektrischrainbow@gmail.com>
Date: Tue, 27 Oct 2026 11:22:04 +0100
Subject: [PATCH 1/3] Find the engine in the MSYS2 layout

MSYS2 installs castle-engine.exe and castle-editor.exe in <prefix>/bin
and the engine (src/, packages/, tools/*/data, ...) in
<prefix>/lib/castle-engine.  Neither $CASTLE_ENGINE_PATH nor the
../ and ../../ of the executable find it there, so the build tool and
the editor start without engine sources and data.

On Windows, also try <exe dir>/../lib/castle-engine in
GetCastleEnginePathSystemWide, after all existing detection methods
(the Unix branch checks /usr/src and /usr/local/src at this point).
---
 tools/common-code/toolcommonutils.pas | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/tools/common-code/toolcommonutils.pas b/tools/common-code/toolcommonutils.pas
index d6e832b..eebb19f 100644
--- a/tools/common-code/toolcommonutils.pas
+++ b/tools/common-code/toolcommonutils.pas
@@ -334,6 +334,22 @@ begin
     Exit;
   {$endif}
 
+  {$ifdef MSWINDOWS}
+  { MSYS2: tools are installed in <prefix>/bin, the engine in
+    <prefix>/lib/castle-engine. }
+  try
+    // knowingly using deprecated ExeName, like GetCastleEnginePathFromExeName
+    {$warnings off}
+    Result := InclPathDelim(ExtractFileDir(ExtractFileDir(ExeName))) +
+      'lib' + PathDelim + 'castle-engine' + PathDelim;
+    {$warnings on}
+    if CheckCastlePath(Result) then
+      Exit;
+  except
+    on EExeNameNotAvailable do ;
+  end;
+  {$endif}
+
   Result := '';
 end;
 
-- 
2.55.0

